source/extensions/tracers/opentelemetry/resource_detectors/environment/environment_resource_detector.cc

```diff
@@ -0,0 +1,56 @@
+#include "environment_resource_detector.h"
+
+#include <sstream>
+#include <string>
+
+#include "source/common/config/datasource.h"
+#include "source/extensions/tracers/opentelemetry/resource_detectors/resource_detector.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace OpenTelemetry {
+
+constexpr absl::string_view kOtelResourceAttributesEnv = "OTEL_RESOURCE_ATTRIBUTES";
+
+/**
+ * @brief Detects a resource from the OTEL_RESOURCE_ATTRIBUTES environment variable
+ * Based on the OTel C++ SDK:
+ * https://github.com/open-telemetry/opentelemetry-cpp/blob/v1.11.0/sdk/src/resource/resource_detector.cc
+ *
+ * @return Resource A resource with the attributes from the OTEL_RESOURCE_ATTRIBUTES environment
+ * variable.
+ */
+Resource EnvironmentResourceDetector::detect() {
+  envoy::config::core::v3::DataSource ds;
+  ds.set_environment_variable(kOtelResourceAttributesEnv);
+
+  Resource resource;
+  resource.schemaUrl = "";
+
+  TRY_NEEDS_AUDIT {
+    auto attributes_str = Config::DataSource::read(ds, true, context_.serverFactoryContext().api());
+    if (attributes_str.empty()) {
+      return resource;
+    }
+
+    std::istringstream iss(attributes_str);
+    std::string token;
+    while (std::getline(iss, token, ',')) {
+      size_t pos = token.find('=');
+      std::string key = token.substr(0, pos);
+      std::string value = token.substr(pos + 1);
+      resource.attributes[key] = value;
+    }
+  }
+  END_TRY catch (const EnvoyException& e) {
+    ENVOY_LOG(error, "Failed to read resource attributes: {}.", e.what());
+  }
+
+  return resource;
```

Comment:
I wanted to not fail and stop Envoy from starting because today, we only load the "resource" with `service.name`. If now I change it to stop initialization I feel this would be a breaking change in behavior somehow.  

This is because there might be cases where a failure in enrichment is OK and not a "stop the world" situation. I will give an example from Dynatrace. Depending where/how the application is deployed, our agent "put" certain files with a bunch of attributes it gathers. The Dynatrace detector then tries to read from all of these files, since it does not know where the app is deployed (for ex in K8s). So it's expected that one of the files will not be present, if the app is say, not running in k8s.

Another example for this detector is maybe the env variable is not set (or has a typo). I don't think this should cause Envoy to exit. We can at maximum log a warning that this happened.

Additionally, in the OTel collector, all of the detectors return an empty resource if they encounter an error:

- [Azure](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/internal/azure/azure.go#L49)
- [Environment](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/internal/env/env.go#L58)
- [System](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/internal/system/system.go#L64)
- [Docker](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/internal/docker/docker.go#L52)

Reply:
>OOC, and apologies for my lack of OTel understanding

First, not a problem at all! You also help me on my lack of Envoy understanding :D 

> but if there's only a warning somewhere, due to misconfiguration, then:
> 
>  1. does it require an operator to look at the logs of of all deployed K8s to see which succeeded and which failed?
> 

That would be one option, but the spans are exported to a back-end and it will be obvious that the service is missing resource attributes, as they are an "essential part" in defining the service. So, I imagine if, for example, they don't set the environment variable where the attributes are, and when they are looking at spans/querying they won't find it for N services and that will be the indication that something is not right.

>  2. If an error occurs, is a default "Resource" is used or is the previously configured "Resource" is used?

Not sure I follow, but the Tracer/Driver is configured when Envoy starts up and it's not changed at runtime IIUC right? 
If that's true, then when a detector fails to "detect", an empty resource is returned, which is then can be merged with other resources detectors the user has configured. 

For ex, I have resource detector A configured and working. Then I configured another resource detector B. When envoy starts up, detector A works but B fails. Envoy will then start with resource A only.

For this specific resource detector that looks at the environment variable, the OTel C++ SDK does also return empty, for ex if the env variable is not set https://github.com/open-telemetry/opentelemetry-cpp/blob/ad626cec4598165e0d297b0bdb3880968c5dfe88/sdk/src/resource/resource_detector.cc.

Reply:
Thanks for explaining this. I've got a few followup questions here:

> > but if there's only a warning somewhere, due to misconfiguration, then:
> > 
> > 1. does it require an operator to look at the logs of of all deployed K8s to see which succeeded and which failed?
> 
> That would be one option, but the spans are exported to a back-end and it will be obvious that the service is missing resource attributes, as they are an "essential part" in defining the service. So, I imagine if, for example, they don't set the environment variable where the attributes are, and when they are looking at spans/querying they won't find it for N services and that will be the indication that something is not right.
> 
I'm assuming there will be cases where Envoy is operated by one entity, and the target servers that receive the spans will be operated by another entity. In that case, one may send an invalid configuration to Envoy, not knowing that it was invalid, and not receive the expected behavior in the target servers.
If the config is accepted, when trying to answer the question "where is the problem?", it could either be the component that sends Envoy the config, or in Envoy, or on the way back to the target servers. One will need to look at the logs of all these, in order to understand where the problem is.
If, however, the config is rejected, this should raise an issue, because the config-server entity will receive a NACK.

> > 2. If an error occurs, is a default "Resource" is used or is the previously configured "Resource" is used?
> 
> Not sure I follow, but the Tracer/Driver is configured when Envoy starts up and it's not changed at runtime IIUC right? If that's true, then when a detector fails to "detect", an empty resource is returned, which then can be merged with other resources detectors the user has configured.
> 
I think that a tracer/driver can be updated dynamically, not only during startup.
Envoy allows for dynamic config update using the [xds protocol](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#xds-rest-and-grpc-protocol).

> For ex, I have resource detector A configured and working. Then I configured another resource detector B. When envoy starts up, detector A works but B fails. Envoy will then start with resource A only.
> 
> For this specific resource detector that looks at the environment variable, the OTel C++ SDK does also return empty, for ex if the env variable is not set https://github.com/open-telemetry/opentelemetry-cpp/blob/ad626cec4598165e0d297b0bdb3880968c5dfe88/sdk/src/resource/resource_detector.cc.

Reply:
@adisuissa  I understand your points. I guess we just need to decide in which approach we want to adopt here. 

As I explained, some resource detectors may "fail" (or be unable) to detect attributes and that is not to be considered an error (the case with Dynatrace I explained is one example). For this case, the spec is clear: (emphasis on this part)

> Errors should be handled as specified in the [Error Handling principles](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.24.0/specification/error-handling.md#basic-error-handling-principles). **Note the failure to detect any resource information MUST NOT be considered an error**, whereas an error that occurs during an attempt to detect resource information SHOULD be considered an error.

But the [Error Handling principles](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.24.0/specification/error-handling.md#basic-error-handling-principles) says:

>The API or SDK MAY fail fast and cause the application to fail on initialization, e.g. because of a bad user config or environment, but MUST NOT cause the application to fail later at run time, e.g. due to dynamic config settings received from the Collector.

Which aligns with what you are saying about Envoy - bad configuration at bootstrap of Envoy fails but at dynamic changes won't stop it. 

For the environment resource detector, if you think it aligns best with Envoy's practices, I can make it fail, for example if the resource detector was configured, but the Env variable is not present OR if the format of the data in the env variable is wrong. But I think this decision to "fail or not and stop Envoy" may be changing depending of the "error" or situation a resource detector has.

Reply:
Ok! I will change it to reject the config if:

- Env variable is not present
- Env variable is present but empty
- ( will try if possible) Env variable is present but with values with "wrong" format

Reply:
Done in [d20885e](https://github.com/envoyproxy/envoy/pull/29547/commits/d20885ea2dc178424a9b7ed30c61240beb78591d)

Now if the cases above happen, Envoy will not start.